linkd-lib/src/protocol.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright © 2021 The Radicle Link Contributors
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::{net::SocketAddr, panic, time::Duration};
+
+use futures::{future::FutureExt as _, pin_mut, select};
+use tokio::{sync::mpsc, time::sleep};
+use tracing::{error, info, instrument};
+
+use librad::{
+    net::{self, discovery::Discovery, peer::Peer},
+    Signer,
+};
+
+#[instrument(name = "peer subroutine", skip(disco, peer, shutdown_rx))]
+pub async fn routine<D, S>(
+    peer: Peer<S>,
+    disco: D,
+    mut shutdown_rx: mpsc::Receiver<()>,
+) -> anyhow::Result<()>
+where
+    D: Discovery<Addr = SocketAddr> + Clone + 'static,
+    S: Signer + Clone,
+{
+    let shutdown = shutdown_rx.recv().fuse();
+    futures::pin_mut!(shutdown);
+
+    loop {
+        match peer.bind().await {
+            Ok(bound) => {
+                let (stop, run) = bound.accept(disco.clone().discover());
+                let run = run.fuse();
+                pin_mut!(run);
```

Comment:
I hate that this is a thing

Reply:
:mage_man: :sparkles:

Reply:
Unfortunately fixing async Rusts Pinning abstraction is out-of-scope for this PR. :<